compal/__init__.py

```diff
@@ -672,30 +696,206 @@ def transform_radio(radio_settings):  # rs = radio_settings
+            out_s.append(('wlBandMode', settings.band_mode))  # change
 
         for item_2g, item_5g in zip(
-            transform_radio(settings.radio_2g).items(),
-            transform_radio(settings.radio_5g).items(),
+                transform_radio(settings.radio_2g),
+                transform_radio(settings.radio_5g),
         ):
             out_s.append(item_2g)
             out_s.append(item_5g)
 
-            if item_2g[0] == "wlHiden5g":
+            if item_5g[0] == "wlHiden5g":
                 out_s.append(("wlCoexistence", settings.bss_coexistence))
 
+        if setter_code == SetFunction.WIFI_SIGNAL:
+            out_s.append(('wlSmartWiFi', settings.smart_wifi))  # change
         # Join the settings
         out_settings = OrderedDict(out_s)
 
-        return self.modem.xml_setter(SetFunction.WIFI_SETTINGS, out_settings)
+        if debug:
+            print(f"\nThe following variables will be sent over 'fun:{setter_code}'"
+                  f" to the router for settting it:\n" + str(out_settings))
+
+        return self.modem.xml_setter(setter_code, out_settings)  # change
+
+    def print_xml_content(self, point=""):
+        """
+        Print out xml-entries in the router that contain wifi-settings
+        """
+        print(f"\n--- SETTINGS {point} ---:")
+        # WIFI State
+        xml_content = self.modem.xml_getter(315, {}).content
+        print("\n ------------------------- WIRELESSBASIC_2 (315) IN ROUTER: : -------------------------\n"
+              + xml_content.decode('utf8'))
+        xml_content = self.modem.xml_getter(326, {}).content
+        print("\n ------------------------- WIFISTATE (326) IN ROUTER: : -------------------------\n"
+              + xml_content.decode('utf8'))
+        xml_content = self.modem.xml_getter(300, {}).content
+        print("\n ------------------------- WIRELESSBASIC (300) IN ROUTER: : -------------------------\n"
+              + xml_content.decode('utf8'))
+        print(str(self.wifi_settings))
+        time.sleep(1)
+
+    @staticmethod
+    def __compare_wifi_settings(old_settings, new_settings):
+        """
+        Compare two settings objects for changes and
+        return:
+            bln_changes: True if there were changes
+            changes: Python dict that contains the changes
+        """
+
+        def iterate_changes(old_settings, new_settings, changes):
+            for attr, old_value in old_settings.__dict__.items():
+                if isinstance(old_value, BandSetting):
+                    changes.update({attr: {}})
+                    iterate_changes(getattr(old_settings, attr), getattr(new_settings, attr), changes[attr])
+                else:
+                    new_value = getattr(new_settings, attr)
+                    if old_value != new_value:
+                        changes.update({attr: f"{old_value} -> {new_value}"})
+                        bln_changes[0] = True
+
+        changes = {}
+        bln_changes = [False]
```

Comment:
It may be possible to do this with a regular boolean instead of the boolean array.

Reply:
I was trying. But with the subprocedure it never worked. When i called iterate_changes inside of interate_changes it always overritten the variable. I tried also with the 'global' keyword. Only the object method in form of a list worked to reference it. Perhaps you find a way. I tried some time.

Reply:
A bit late, but: I think the def captures a reference to the variable in the scope it is defined in. I don't know how to work around this in a clean fashion.

In other languages you could explicit object references (AtomicReference in Java). In python perhaps you can use a threadlocal (https://docs.python.org/3.8/library/threading.html#thread-local-data).

Alternatively, when you need to define functions in a function, it may be a sign that it is time to create a class.